### PR TITLE
Fix some pipeline tests

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

- change `feature_extractor=...` to `image_processor=...` in `test_pipeline_XXX` files, if `XXX` is vision tasks.
  - this doesn't affect any backward compatibility - they are just testing files
- change `self.feature_extractor(...)` to `self.image_processor()` in some vision pipeline files
  - the backward compatibility is ensured by the change in `base.py` (parent class `Pipeline.__init__`):
    ```python
    if self.image_processor is None and self.feature_extractor is not None:
        # Backward compatible change, if users called
         # ImageSegmentationPipeline(.., feature_extractor=MyFeatureExtractor())
         # then we should keep working
         self.image_processor = self.feature_extractor
    ```
- A few other fixes, see review comments

### CODE REVIEW

File: tests/pipelines/test_pipelines_document_question_answering.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ class DocumentQuestionAnsweringPipelineTests(unittest.TestCase, metaclass=Pipeli
     @require_vision
     def get_test_pipeline(self, model, tokenizer, processor):
         dqa_pipeline = pipeline(
-            "document-question-answering", model=model, tokenizer=tokenizer, feature_extractor=processor
+            "document-question-answering", model=model, tokenizer=tokenizer, image_processor=processor
```
Comment (ydshieh) [inline tests/pipelines/test_pipelines_document_question_answering.py:64]: better + no backward compatibility issue (just testing files)

File: src/transformers/pipelines/depth_estimation.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def _sanitize_parameters(self, **kwargs):
     def preprocess(self, image):
         image = load_image(image)
         self.image_size = image.size
-        model_inputs = self.feature_extractor(images=image, return_tensors=self.framework)
+        model_inputs = self.image_processor(images=image, return_tensors=self.framework)
```
Comment (ydshieh) [inline src/transformers/pipelines/depth_estimation.py:90]: backward compatibility is ensured by the change (of `__init__`) in `base.py`, see comment there.

File: src/transformers/pipelines/base.py [hunk 3]
```diff
@@ -792,6 +792,12 @@ def __init__(
         self._num_workers = kwargs.pop("num_workers", None)
         self._preprocess_params, self._forward_params, self._postprocess_params = self._sanitize_parameters(**kwargs)
 
+        if self.image_processor is None and self.feature_extractor is not None:
```
Comment (ydshieh) [inline src/transformers/pipelines/base.py:795]: To ensure backward compatibility

File: src/transformers/pipelines/base.py [hunk 4]
```diff
@@ -77,7 +77,7 @@ def _pad(items, key, padding_value, padding_side):
         # Others include `attention_mask` etc...
         shape = items[0][key].shape
         dim = len(shape)
-        if key == "pixel_values":
+        if key in ["pixel_values", "image"]:
```
Comment (ydshieh) [inline src/transformers/pipelines/base.py:80]: `DocumentQuestionAnsweringPipelineTests.get_test_pipeline` prepare inputs with "`image` key.

- not sure why not `pixel_values` there.
-  `DocumentQuestionAnsweringPipelineTests` failed after #20426, because of the usage of `LayoutLMv2`, The tests with this model were all skipped previously (i.e. before #20426)

File: src/transformers/pipelines/__init__.py [hunk 5]
```diff
@@ -387,8 +387,11 @@
     if values["type"] == "text":
         NO_FEATURE_EXTRACTOR_TASKS.add(task)
         NO_IMAGE_PROCESSOR_TASKS.add(task)
-    elif values["type"] in {"audio", "image", "video"}:
+    elif values["type"] in {"image", "video"}:
         NO_TOKENIZER_TASKS.add(task)
+    elif values["type"] in {"audio"}:
+        NO_TOKENIZER_TASKS.add(task)
+        NO_IMAGE_PROCESSOR_TASKS.add(task)
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:394]: to avoid `AutomaticSpeechRecognitionPipeline` to load an image processor (with the new addition below in this file)

File: src/transformers/pipelines/__init__.py [hunk 6]
```diff
@@ -773,6 +776,10 @@ def pipeline(
     load_feature_extractor = type(model_config) in FEATURE_EXTRACTOR_MAPPING or feature_extractor is not None
     load_image_processor = type(model_config) in IMAGE_PROCESSOR_MAPPING or image_processor is not None
 
+    # Just keep `load_image_processor`
+    if load_image_processor and load_feature_extractor:
+        load_feature_extractor = False
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:781]: This is arguably a bit breaking, if one want to access `self.feature_extractor`. I can remove this block if @sgugger also agree it is better/should.

File: src/transformers/pipelines/__init__.py [hunk 7]
```diff
@@ -784,6 +791,18 @@ def pipeline(
         # so the model_config might not define a tokenizer, but it seems to be
         # necessary for the task, so we're force-trying to load it.
         load_tokenizer = True
+    if (
```
Comment (ydshieh) [inline src/transformers/pipelines/__init__.py:794]: To solve some issues, but in generally, we should have this block too, just as for tokenizer/feature_extractor above and below.

**Review by ydshieh — COMMENTED** [review]
Left a comments to facilitate review process.

File: src/transformers/pipelines/__init__.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/pipelines/__init__.py:781]: Mmm what happens for a model that mixes image and sound then?

File: tests/pipelines/test_pipelines_common.py [hunk 8]
```diff
@@ -179,6 +179,18 @@ def is_test_to_skip(test_casse_name, config_class, model_architecture, tokenizer
...
+            "TextClassificationPipelineTests",
+            "TokenClassificationPipelineTests",
+            "ZeroShotClassificationPipelineTests",
+        ]:
+            # `LayoutLMv2Config` was never used in pipeline tests (`test_pt_LayoutLMv2Config_XXX`) due to lack of tiny
+            # config. With new tiny model creation, it is available, but we need to fix the failed tests.
+            to_skip = True
+        elif test_casse_name == "DocumentQuestionAnsweringPipelineTests" and not tokenizer_name.endswith("Fast"):
+            # This pipeline uses `sequence_ids()` which is only available for fast tokenizers.
+            to_skip = True
```
Comment (sgugger) [inline tests/pipelines/test_pipelines_common.py:193]: We should not test LayoutLMv2 at all then, until we have the common tester for pipelines and we can overwrite the common tests in individual test files. The common test should not contain a list of tests on configs/test names.

**Review by sgugger — COMMENTED** [review]
Would like to also have @Narsil have a look at this.

Let's stop adding tests of new models in the pipelines until the metaclass is removed and we have a mixin class that all model testers inherit from (I think that is the plan, right?) as we can't add a succession of tests in the pipeline common tests to change behavior for this or that model.

File: src/transformers/pipelines/__init__.py (hunk 6 above)
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:781]: Yeah, good question. I don't think we have this use case yet, but as I also have a doubt mentioned above, I can simply remove this block to avoid any potentially issue.

File: tests/pipelines/test_pipelines_common.py (hunk 8 above)
Comment (ydshieh) [reply tests/pipelines/test_pipelines_common.py:193]: Agree! This big `is_test_to_skip`  is added in #20426 to server as a record + to pass CI in order to merge.

**Comment by ydshieh** [comment]
> Let's stop adding tests of new models in the pipelines until the metaclass is removed and we have a mixin class that all model testers inherit from (I think that is the plan, right?) 

Yes, that's the plan! For now, no new models will be added in pipeline testing (because we need to create and upload the tiny model on the Hub, and I guess no one knows how to do it except me, and I definitely will keep my life easier :-)

> as we can't add a succession of tests in the pipeline common tests to change behavior for this or that model.
Sure!

File: tests/pipelines/test_pipelines_common.py (hunk 8 above)
Comment (ydshieh) [reply tests/pipelines/test_pipelines_common.py:193]: (BTW, `LayoutLMv2`  was never used in pipeline testing before #20426, but after that PR, it works in `DocumentQuestionAnsweringPipelineTests`)

File: src/transformers/pipelines/__init__.py (hunk 6 above)
Comment (Narsil) [reply src/transformers/pipelines/__init__.py:781]: Yes let's remove this and keep this as simple as possible (it's already quite tortuous)

File: tests/pipelines/test_pipelines_document_question_answering.py [hunk 9]
```diff
@@ -82,7 +82,8 @@ def get_test_pipeline(self, model, tokenizer, processor):
                 "word_boxes": word_boxes,
             },
             {
-                "image": None,
+                # if image processor is available, we always need to prepare `image`, otherwise let's test without it
+                "image": load_image(image) if processor is not None else None,
```
Comment (Narsil) [inline tests/pipelines/test_pipelines_document_question_answering.py:86]: I don't understand this.

Let's just remove this test case, `image:None ` in a document-qa pipeline doesn't  make a whole lot of sense in general imo.

File: tests/pipelines/test_pipelines_common.py (hunk 8 above)
Comment (Narsil) [reply tests/pipelines/test_pipelines_common.py:193]: I agree with @sgugger I think,

Let's escape individually `layoutlmv2` in all the pipeline it's attempted to be called.
IT makes it more obvious to readers that the model is broken, and we can enable them one by one.

**Review by Narsil — APPROVED** [review]
LGTM

File: tests/pipelines/test_pipelines_common.py (hunk 8 above)
Comment (ydshieh) [reply tests/pipelines/test_pipelines_common.py:193]: I am going to start implementing common pipeline testing (as common model testing), and the list here is not only to skip tests, but more importantly to record what tests need to be fixed.

As it is just temporarily, and `is_test_to_skip` already contains lot of skip cases, I don't think it hurts (more) to add one more case (and again, at least I will know clearly what tests should be fixed or to be overwritten in the common pipeline testings PR)

File: src/transformers/pipelines/__init__.py (hunk 6 above)
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:781]: OK

File: src/transformers/pipelines/__init__.py [hunk 10]
```diff
@@ -784,6 +791,18 @@ def pipeline(
...
+        image_processor is None
+        and not load_image_processor
+        and normalized_task not in NO_IMAGE_PROCESSOR_TASKS
+        # Using class name to avoid importing the real class.
+        and model_config.__class__.__name__ in MULTI_MODEL_CONFIGS
+        and normalized_task != "automatic-speech-recognition"
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:800]: Without this `normalized_task != "automatic-speech-recognition"`, the flag `load_image_processor = True` is set, and `AutomaticSpeechRecognitionPipeline` will try to use auto image processor class to load a speech model checkpoint on the Hub - and fails.

[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:800]: We can probably refine `MULTI_MODEL_CONFIGS` to be more fine granularity, as it currently doesn't distinguish text-vision or text-speech cases, and leads to such issue mentioned above

File: src/transformers/pipelines/base.py (hunk 3 above)
Comment (ydshieh) [reply src/transformers/pipelines/base.py:795]: You are right, I am so ignorant , sorry. I guess it won't hurt, but potentially leads to some confusion if one abuse the attributes.

One way is to verify if `self.feature_extractor` is an instance of `BaseImageProcessor`

File: src/transformers/pipelines/document_question_answering.py [hunk 11]
```diff
@@ -281,7 +281,9 @@ def preprocess(
         image_features = {}
         if input.get("image", None) is not None:
             image = load_image(input["image"])
-            if self.feature_extractor is not None:
+            if self.image_processor is not None:
+                image_features.update(self.image_processor(images=image, return_tensors=self.framework))
+            elif self.feature_extractor is not None:
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/pipelines/document_question_answering.py:286]: Yeah, let me check. It's probably the ordering of the fixed I have done - I did this block first, and needed to change `base.py` to fix other failures ...

File: tests/pipelines/test_pipelines_document_question_answering.py (hunk 9 above)
Comment (ydshieh) [reply tests/pipelines/test_pipelines_document_question_answering.py:86]: OK! Just want to pointing out that when this pipeline test is run with `LayoutLM` model (not `v2`), it works fine, as it seems there is no image processor associated in this case.

**That's why I am not sure if this document-qa pipeline should or not (or both)  contain the image inputs.**

File: src/transformers/pipelines/document_question_answering.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/pipelines/document_question_answering.py:286]: well, this is needed if we still want to keep `LayoutLM` for this pipeline testing.

File: tests/pipelines/test_pipelines_document_question_answering.py (hunk 9 above)
Comment (Narsil) [reply tests/pipelines/test_pipelines_document_question_answering.py:86]: > That's why I am not sure if this document-qa pipeline should or not (or both) contain the image inputs.

I think support was added because of LayoutLM. Strange (and sad) that V2 and V3 were never really tested.

Personally I would simply remove the `None` from the test, it's too niche to be valuable, imo. (What kind of document question answering, answers without a document ! :D That's called plain question-answering :) )

Comment (ydshieh) [reply tests/pipelines/test_pipelines_document_question_answering.py:86]: OK

**Comment by ydshieh** [comment]
Failing test is irrelevant to this PR.

**Review by sgugger — APPROVED** [review]
Good for me, but as I said before, let's make the common pipeline tester first before adding new models :-)

**Comment by ydshieh** [comment]
For transparancy: I need to add back the following block in `src/transformers/pipelines/__init__.py`:
```python
    # If `model` (instance of `PretrainedModel` instead of `str`) is passed (and/or same for config), while
    # `image_processor` or `feature_extractor` is `None`, the loading will fail. This happens particularly for some
    # vision tasks when calling `pipeline()` with `model` and only one of the `image_processor` and `feature_extractor`.
    # TODO: we need to make `NO_IMAGE_PROCESSOR_TASKS` and `NO_FEATURE_EXTRACTOR_TASKS` more robust to avoid such issue.
    # This block is only temporarily to make CI green.
    if load_image_processor and load_feature_extractor:
        load_feature_extractor = False
```

 The issue comes from the fact of calling `pipeline()` with:

 - a model (not string)
 - one of `image_processor` or `feature_extractor` being specified, but another one is `None`
 - tasks involving vision models, so both `load_image_processor` and `load_feature_extractor` are `True`

then it will fail around https://github.com/huggingface/transformers/blob/c2f623cf53a9a8b2e192135b03ae211ba1ce3695/src/transformers/pipelines/__init__.py#L863

Without this change, the following 2 tests just fails
- DocumentQuestionAnsweringPipelineTests::test_pt_LayoutLMv2Config_LayoutLMv2ForQuestionAnswering_LayoutLMv2TokenizerFast_LayoutLMv2ImageProcessor (not tested before)
- tests/pipelines/test_pipelines_image_segmentation.py::ImageSegmentationPipelineTests::test_maskformer (already failed since the PR #20851 one week ago)

**We need to improve this `pipeline.__init__.py` to make it more robust regarding the feature_extractor/image_processor while we want to keep backward compatibility**

This should go in a separate PR though, I will merge this PR as it is, unless you strongly against the changes in the last 2 commits.
